### PR TITLE
Give UDP connections a chance to close gracefully

### PR DESCRIPTION
**Pull request by jsravn**
**What this PR does / why we need it**:

Delay UDP connection tracking flush on endpoint removal.

This gives termination grace period a chance to work, such as when
kube-dns is restarted or redeployed.

Otherwise in-flight requests will always time out whenever an endpoint
removal occurs.

We get DNS timeouts occasionally when kube-dns redeploys. I believe the immediate UDP flush is what causes it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
For #45976

**Special notes for your reviewer**:

**Release note**:
```release-note
Delay UDP connection flush to give connections a chance to gracefully terminate.
```

### CODE REVIEW

**Comment by krmayankk** [comment]
@jsravn  could we delay the udp connection flush by some other param like terminationGracePeriodSeconds rather than hard coding it ? May be some users want that, immediate flush ?

**Comment by jsravn** [comment]
@krmayankk I added it as a flag to kube-proxy so it's configurable - setting to 0 disables, same as prior behavior. It is also similar to TCP connections in netfilter, they have a single timeout.

I did think about using terminationGracePeriodSeconds as described in the linked issue. The problem is kube-proxy only knows about endpoints - and terminationGracePeriodSeconds is a pod field. It's doable but would require an API change probably - adding a new field to endpoint that is filled in by the endpoints controller. I think this is a good idea actually, just a bit more work and API change (which I'm not that familiar with doing).

**Comment by jsravn** [comment]
Using terminationGracePeriod also makes sense for the IPVS proxier (https://github.com/kubernetes/kubernetes/issues/57841).

**Review by jhorwit2 — CHANGES_REQUESTED** [review]

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -673,15 +676,37 @@ func servicePortEndpointChainName(servicePortName string, protocol string, endpo
...
+				proxier.deleteEndpointPair(svcInfo, epSvcPair)
+			} else {
+				// Delay UDP flush so in-flight requests are not immediately dropped before they have a chance to complete.
+				go func() {
+					time.Sleep(proxier.udpConnectionFlushDelay)
+					proxier.mu.Lock()
+					defer proxier.mu.Unlock()
+					if endpoints, ok := proxier.endpointsMap[epSvcPair.ServicePortName]; ok {
+						for _, ep := range endpoints {
+							if ep.String() == epSvcPair.Endpoint {
```
[the remark this answers is not in the document]
Comment (jsravn) [reply pkg/proxy/iptables/proxier.go]: Fixed.

**Comment by bburket** [comment]
@jsravn how did your testing go? We are facing some cantankerous issues with DNS timeouts in our cluster

**Comment by jsravn** [comment]
@bburket Not great I'm afraid. I still am getting dns errors on restarts, with this change cherry picked to 1.8. Either my cherry pick is broken or there is something else going on I don't quite understand yet.

**Comment by tcolgate** [comment]
We are seeing similar issues with kube-dns on cluster-autoscaler downscales. NodeJS seems to have particular pathological behaviour on DNS timeouts, so this is causing us some considerable pain.

**Comment by jsravn** [comment]
Sorry I got pulled off on other work, so haven't had time to work on this further.

**Comment by m1093782566** [comment]
/assign 

Probably IPVS side needs a similar change.

**Comment by BenTheElder** [comment]
@m1093782566 seems like we probably still want this...?
/remove-lifecycle stale

**Comment by m1093782566** [comment]
@BenTheElder 

Thanks but I would prefer #66012 which is up-to-date and fix both TCP and UDP issues.

**Comment by BenTheElder** [comment]
It looks like https://github.com/kubernetes/kubernetes/pull/66012 only contains IPVS changes though, while this one has iptables?

**Comment by m1093782566** [comment]
oops, seems you are right. This PR covers iptables while  #66012 covers IPVS - they do the different thing.

**Comment by jsravn** [comment]
Going to close this - as a lot of behavior currently relies on the UDP connections being dropped immediately. IPVS proxier has also done the same recently.

**Pull request closed without merging**